internal/pipeline/sign.go

```diff
@@ -0,0 +1,186 @@
+package pipeline
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"hash"
+	"io"
+	"sort"
+)
+
+type newSignatureScheme = func([]byte) (signatureScheme, error)
+
+var algorithms = map[string]newSignatureScheme{
+	"hmac-sha256": newHMACSHA256,
+}
+
+// signatureScheme describes what a signature scheme needs to do to support
+// the Sign func and Signature.Verify method.
+type signatureScheme interface {
+	io.Writer
+
+	// Sign returns a signature for the data written so far.
+	Sign() ([]byte, error)
+
+	// Verify checks a given signature is valid for the data written so far.
+	Verify([]byte) error
+}
+
+type hmacSHA256 struct {
+	hash hash.Hash
+}
+
+func newHMACSHA256(key []byte) (signatureScheme, error) {
+	return hmacSHA256{
+		hash: hmac.New(sha256.New, key),
+	}, nil
+}
+
+func (h hmacSHA256) Write(b []byte) (int, error) {
+	return h.hash.Write(b)
+}
+
+func (h hmacSHA256) Sign() ([]byte, error) {
+	return h.hash.Sum(nil), nil
+}
+
+func (h hmacSHA256) Verify(sig []byte) error {
+	c := h.hash.Sum(nil)
+	if !bytes.Equal(c, sig) {
+		return fmt.Errorf("message digest mismatch (%x != %x)", c, sig)
+	}
+	return nil
+}
+
+// Signature models a signature (on a step, etc).
+type Signature struct {
+	Algorithm    string   `json:"algorithm" yaml:"algorithm"`
+	SignedFields []string `json:"signed_fields" yaml:"signed_fields"`
+	Value        string   `json:"value" yaml:"value"`
+}
+
+// signedFielder describes types that can be signed and have signatures
+// verified.
+// Converting non-string fields into strings (in a stable, canonical way) is an
+// exercise left to the implementer.
+type signedFielder interface {
+	// signedFields returns the default set of fields to sign, and their values.
+	// This is called by Sign.
+	signedFields() map[string]string
+
+	// valuesForFields looks up each field and produces a map of values. This is
+	// called by Verify. The set of fields might differ from the default, e.g.
+	// when verifying older signatures computed with fewer fields or deprecated
+	// field names. signedFielder implementations should reject requests for
+	// values if "mandatory" fields are missing (e.g. signing a command step
+	// should always sign the command).
+	valuesForFields([]string) (map[string]string, error)
+}
+
+// Sign computes a new signature for a signedFielder.
+func Sign(sf signedFielder, algName string, key []byte) (*Signature, error) {
```

Comment:
i wonder if it would make more sense to have something like
```suggestion
// Sign computes a new signature for a signedFielder.
func Sign(sf signedFielder, scheme SigningScheme, key []byte) (*Signature, error) {
```

and then make the method to create a new signing scheme from an algorithm string public:
```Go
func NewSignatureScheme(algorithm string) (SignatureScheme, error) {
	switch algorithm {
	case "hmac-sha256":
		// ...
	case "rsa-pss":
		// ...
	default:
		return nil, fmt.Errorf("unsupported signing algorithm: %s", algorithm)
	}
}
```

that way we don't have to replicate the behaviour between the `Sign` and `Verify` methods, and leave deciding on the signature scheme to the caller

Reply:
that said though, this is entirely stylistic, and i don't know that it makes a huge difference

Reply:
This has prompted some new thoughts...

Reply:
I've taken the idea a bit further and split SignatureScheme into Signer and Verifier. Even though these are the same thing under the hood, it makes it slightly harder to reuse a signer for verifying.